proposals/p1928.md

```diff
@@ -39,125 +44,142 @@ continuous in memory.
-of arrays contains additional metadata:
+For simplifying usage and raising safety of array's usage, internal
+representation of arrays contains additional metadata:
+
 ```cpp
-  struct {
-    int32_t dimensions;
-    int32_t * dimension_size;
-    void * data;
-  } Array;
+template <typename T, int32_t N> struct Array {
+  std::array<int32_t, N> dimension_sizes;
+  std::unique_ptr<T[]> data;
+};
 ```
-Here, `dimensions` contains information about number of dimensions (necessary
-for multidimensional arrays), `dimension_size` is an array of size `dimensions`
-containing sizes for each dimensions, and `data` is continuous in memory place
-keeping scalar data.
 
-This representation is closer to `std::vector<T>` rather then `T[]`. Then,
+Here, `dimension_sizes` contains information about dimensions' length. `size()`
+property of `dimension_sizes` contains information about number of dimensions.
+`data` is continuous in memory place keeping scalar data.
+
+This representation is closer to `std::array<T>` rather then `T[]`. Then,
 interoperability between C++ and carbon is limited.
 
 One-dimensional and nested arrays have `dimensions = 1`.
 
 ### Declaration
 
-During declaration of arrays, it is more importantly to see type of variable.
-It is the aim of the below presented alternatives of declaration.
+During declaration of arrays, it is more importantly to see type of variable. It
+is the aim of the below presented alternatives of declaration.
+
+Current realization (like in `Rust`):
 
-Current realization:
 ```carbon
   var array_1D: [i32; 4];
   var array_2D: [i32; 4, 10]; // is not supported yet
   var array_nested: [[i32; 10]; 4];
 ```
+
 In the case of nested array, definition of them a bit misunderstood due to
 following indexing is inversed.
 
-Suggested in [#1787](https://github.com/carbon-language/carbon-lang/pull/1787):
-```carbon
-  var array_1D: [4; i32];
-  var array_2D: [4, 10; i32];
-  var array_nested: [4; [10; i32]];
-```
-This definition is deprived of a defect of previous definition, but the type is
-the last one.
-
-Alternatives:
-```carbon
-  var array_1D[4]: i32;
-  var array_2D[4, 10]: i32;
-  var array_nested[4][10]: i32;
-```
-In this alternative, `auto` will work only for types, not for dimensions of
-array.
-
-```carbon
-  var array_1D: i32[4];
-  var array_2D: i32[4, 10];
-  var array_nested: i32[4][10];
-```
-This definition requires contextual information since `i32[4]` is the same as
-`array_1D[1]` so it is necessary to separate type and variable.
+### Declaration like VLA
 
 Also, it is possible to declare array in function with runtime defined size:
+
 ```carbon
   fn f(n: i32) -> i32 {
     var arr: [n; i32];
     return 0;
   }
 ```
 
-Only of size is defined in compile-time and it is less than
+Only if size is defined in compile-time and it is less than
 `MAX_STACK_SIZE_ARRAY`, array is located in stack memory. Other definitions are
 located in heap memory. It is especially important for the last example.
```

Comment:
When the array storage is on the stack, does this type still use the representation shown above, with a pointer to storage outside the object, or does it use a representation like `std::array`, where the storage is located within the object itself?